app/validators/answer_code_validator.py

```diff
@@ -0,0 +1,110 @@
+        duplicates = find_duplicates(self.codes)
+
+        if len(duplicates) > 0:
+            self.add_error(self.DUPLICATE_ANSWER_CODE_FOUND, duplicates=duplicates)
+
+    def validate_missing_answer_id(self):
+        all_answer_ids = list(self.questionnaire_schema.answers_with_context)
+
+        for answer_code_id in self.answer_codes_answer_ids:
+            if answer_code_id not in all_answer_ids:
+                self.add_error(
+                    self.ANSWER_CODE_ANSWER_ID_NOT_FOUND_IN_SCHEMA,
+                    answer_id=answer_code_id,
+                )
+
+    def validate_missing_answer_codes(self):
+        all_answer_ids = list(self.questionnaire_schema.answers_with_context)
+
+        for answer_id in all_answer_ids:
+            if answer_id not in self.answer_codes_answer_ids:
+                self.add_error(self.MISSING_ANSWER_CODE, answer_id=answer_id)
+
+    def validate_answer_codes_at_option_level(self):
+        answers_with_context = self.questionnaire_schema.answers_with_context
+
+        for answer_id in answers_with_context:
+            answer = answers_with_context[answer_id]
+
+            if "options" in answer["answer"]:
+                values = []
+                values.extend(option["value"] for option in answer["answer"]["options"])
+
+                answer_codes_for_options = [
+                    answer_code
+                    for answer_code in self.answer_codes
+                    if answer_code["answer_id"] == answer_id
+                ]
+
+                if len(values) != len(answer_codes_for_options) and all(
+                    "value" in answer_code for answer_code in answer_codes_for_options
+                ):
+                    self.add_error(
+                        self.ANSWER_CODE_MISSING_FOR_ANSWER_OPTIONS,
+                        answer_options=answer["answer"]["options"],
+                        answer_codes_for_options=answer_codes_for_options,
+                    )
+
+                for answer_code in answer_codes_for_options:
+                    if answer_code.get("value") and answer_code["value"] not in values:
+                        self.add_error(
+                            self.INCORRECT_VALUE_FOR_ANSWER_CODE_WITH_ANSWER_OPTIONS,
+                            value=answer_code["value"],
+                            answer_codes_for_options=answer_codes_for_options,
+                        )
+
+                if len(answer_codes_for_options) != 1 and any(
+                    "value" not in answer_code
+                    for answer_code in answer_codes_for_options
+                ):
+                    # Multiple answer codes for options are allowed where options are dynamic
+                    if not answers_with_context[answer_id]["answer"].get(
+                        "dynamic_options"
+                    ):
+                        self.add_error(
+                            self.MORE_THAN_ONE_ANSWER_CODE_SET_FOR_ANSWER_OPTIONS,
```

Comment:
I found this hard to resolve, it is not clear on what is needed to fix the issue. Based on the current logic this is more often true when you forget to define answer_value for some option. It might be better to say, answer code with matching answer_value not found for option. Worth also adding something like "Either all option values must have a code or none of them should".

Reply:
I think the name of this one was causing confusion - I added this to test that multiple answer codes could only be set if answer options were dynamic. Updated the name of this to reflect this 👍 I have also added a new check and test to cover the scenario that is more like the description of this error, where there are two answer codes with the same value but different codes.